[BEAM-1573] Use Kafka serializers instead of coders in KafkaIO

peay opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---



**This is a work in progress, do not merge**.

Modified:
- Use Kafka serializers and deserializers in KafkaIO
- Added helper methods `fromAvro` and `toAvro`, to use serialization based on `AvroCoder`. This is uniform with other IO such as HDFS.
- Moved `CoderBasedKafkaSerializer` out, and added `CoderBaseKafkaDeserializer`. These are used for `toAvro/fromAvro`, and can be useful to port existing code that relies on coder.
- Added `InstantSerializer` and `InstantDeserializer`, as `Instant` is used in some of the tests.
 
Writer lets Kafka handle serialization itself. Reader uses Kafka byte deserializers, and calls the user-provided Kafka deserializer from `advance`. Note that Kafka serializers and deserializers are not themselves `Serializable`. Hence, I've used a `Class<..>` in the `spec` both for read and write.

There is still an issue, though. `Read` still takes **both a deserializer and a coder**. This is because the source must implement `getDefaultOutputCoder`, and I am not sure how to infer it. Having to provide the two is heavy, but I am not sure how to infer the coders in this context. Any thoughts?

cc @rangadi @jkff

jkff commented:
Thanks, please let me know when this is ready for review.

peay commented:
Thoughts on how to handle coders, as described in the PR, would be welcome. In particular -- do we want to have to specify both coder and deserializer in `Read`, or is there a way to infer it?

jkff commented:
Generally, having both setters (serializer and coder) makes sense. For convenience, it would be nice if the user didn't have to specify the coder in simple cases. I think you can try doing that by inferring the coder, if it's not specified explicitly, from the type of the serializer. Here are the steps for that:

* If a serializer is specified but coder is not, you can inspect the serializer's type using reflection and extract the type parameter of its base class.
* When applying the Read.from() transform to an input, you can use input.getPipeline().getCoderRegistry() to try inferring a coder for that type. Then you'll need to pass that coder into the Source object so you can return it from getDefaultOutputCoder() (by the way, I think that method should be nullable but currently isn't - I'll file a JIRA).

jkff commented:
Hi! Any updates on this / anything I can help with?

peay commented:
I got a bit bogged down with other stuff, I'll try to pick it up in the coming days.

peay commented:
Added:
- Infer `Coder`s from the types of the `Deserializer` using the registry
- Do not use explicit `withXxxCoder` in tests when not necessary
- Minimal tests for `Coder` inference

I have not added tests for `fromAvro/toAvro`.

@jkff @rangadi can you review?

I had quite a bit of trouble building and getting CI to work on my fork. I was getting:
- a lot of beam/sdks/java/javadoc/target/collected/org/apache/beam/sdk/transforms/Filter.java:175: error: unmappable character for encoding ASCII` -- added `-encoding utf8` to javadoc in `pom.xml`
- `checkstyle.xml` not found -- I've changed the path in `pom`.xml`.

My branch has a small commit that fixes those two. I am happy to roll it back before we merge.

Most builds on Travis on my fork are still failing with errors such as `Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:1.4.1:enforce (enforce) on project beam-sdks-java-io-hadoop-jdk1.8-tests`. I also have rebased onto the latest `master`. Is it supposed to build cleanly?

peay commented on sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOTest.java:
```diff
@@ -691,6 +710,24 @@ public void testSinkDisplayData() {
+    CoderRegistry registry = new CoderRegistry();
+    registry.registerStandardCoders();
+
+    assertTrue(KafkaIO.getOrInferCoder(registry, LongDeserializer.class, null)
```
A remark here: most of the tests were using `BigEndianLongCoder`, but the registry infers `VarLongCoder` by default for `Long` instead. I do not believe this is an issue, but bear this in mind.

jkff replied:
I think this is fine. It will not silently cause a user's pipeline to differently interpret the data, right? cause we don't automatically infer deserializers from coders, only vice versa.

peay replied:
Yes, I think that is correct.

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -214,18 +251,62 @@
+        Class<T> clazz = (Class<T>) actualType;
+        return coderRegistry.getDefaultCoder(clazz);
+      } catch (CannotProvideCoderException | ClassCastException e) {
+        LOG.warn("Could not infer coder from deserializer type {}: {}",
```
How about simply LOG.warn(....., e);

rangadi replied:
Alternately, let it me thrown. This method does not need to return null.

peay replied:
I have modified this to let `ClassCastException` be thrown, but `CannotProvideCoderException` is checked and I cannot throw it from `expand` or `validate`. I've modified the logging statement to `LOG.warn(....., e);`,

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -214,18 +251,62 @@
+        e.printStackTrace();
+      }
+    } else {
+      LOG.warn("Could not extract deserializer type from {}", deserializer);
```
Flip the if/else branches to reduce nesting?

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -249,16 +330,74 @@
+   * container.
+   */
+  @SuppressWarnings("unchecked")
+  public static <K, V> Read<K, V> fromAvro(Class<K> keyClass, Class<V> valueClass) {
```
This is pretty cool.

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -249,16 +330,74 @@
 
   /**
+   * Creates an uninitialized {@link Read} {@link PTransform}, using a Kafka {@link Deserializer}
+   * based on {@link AvroCoder}. The coder reads Avro data directly without using an Avro object
```
It's probably worth mentioning what guarantees on the data wire format this gives and doesn't give. Is this intended to read only data produced by KafkaIO.toAvro()?

peay replied:
I have tried to clarify a bit. This reads the Avro binary serialization format.

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -249,16 +330,74 @@
 
+  /**
+   * Creates an uninitialized {@link Write} {@link PTransform}, using a Kafka {@link Serializer}
+   * based on {@link AvroCoder}. The coder writes Avro data directly without using an Avro object
```
Ditto

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -441,15 +599,35 @@ public void validate(PBegin input) {
+                "Key coder must be set");
+        checkNotNull(getOrInferCoder(registry, getValueDeserializer(), getValueCoder()),
+                "Value coder must be set");
+      } else {
```
I'd assume that we can hit this only in tests, right? Can we fix the tests to supply a non-null value here? i.e. TestPipeline.begin()?

peay replied:
Yes, I think so, but only some of them, such as `testUnboundedSourceSplits`, which manually calls`generateInitialSplits`  and `Read.from` on each split.  This means that `validate` is called through `Source.validate` which doesn't get a `PInput`

peay replied:
Any advice on how I could change the tests to avoid this extra check? Do we actually want to keep it as is? This is a way the source can be used, after all - even if a bit unusual.

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -441,15 +599,35 @@ public void validate(PBegin input) {
+      // Infer key/value coders if not specified explicitly
+      CoderRegistry registry = input.getPipeline().getCoderRegistry();
+
+      // Handles unbounded source to bounded conversion if maxNumRecords or maxReadTime is set.
```
This comment is misplaced

peay replied:
Fixed.

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -1003,15 +1194,15 @@ public boolean advance() throws IOException {
           curRecord = null; // user coders below might throw.
 
-          // apply user coders. might want to allow skipping records that fail to decode.
+          // apply user coders. might want to allow skipping records that fail to deserialize.
```
s/coders/deserializers

peay replied:
Fixed.

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -1253,26 +1444,21 @@ public void validate(PCollection<KV<K, V>> input) {
       checkNotNull(getTopic(), "Kafka topic should be set");
-      checkNotNull(getKeyCoder(), "Key coder should be set");
-      checkNotNull(getValueCoder(), "Value coder should be set");
+      //checkNotNull(getKeyCoder(), "Key coder should be set");
```
What's up with this?

peay replied:
Fixed, removed.

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/serialization/InstantDeserializer.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Override
+    public Instant deserialize(String topic, byte[] bytes) {
+        return new Instant(longDeserializer.deserialize(topic, bytes) + Long.MIN_VALUE);
```
Add a comment why adding MIN_VALUE?

peay replied:
I have no idea :-). Fixed: `Instant{Serializer,Deserializer}` now just serializes the long value.

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/serialization/InstantSerializer.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Override
+    public byte[] serialize(String topic, Instant instant) {
+        return longSerializer.serialize(topic, instant.getMillis() - Long.MIN_VALUE);
```
Subtracting Long.MIN_VALUE seems fishy.

peay replied:
Good points on this instant serialization business. Some of the tests were using coder-based serialization with `InstantCoder`, and I duplicated the Coder's encoding in there to avoid changing too much, which is not really a good reason.

I'll change this serializer to remove all references to `InstantCoder` and simply serialize `instant.getMillis()` as is.

peay replied:
Fixed.

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/serialization/InstantDeserializer.java:
```diff
@@ -0,0 +1,51 @@
+/**
+ * Kafka {@link Deserializer} for {@link Instant}.
+ *
+ * <p>This uses the same encoding as {@link org.apache.beam.sdk.coders.InstantCoder}.
```
Why? It'd be better to document the wire format used by this Deserializer explicitly, or via LongDeserializer ("deserializes the Instant as millis since epoch using LongDeserializer"), since the format of InstantCoder is not guaranteed to be anything in particular.

peay replied:
Fixed: `Instant{Serializer,Deserializer}` now just serializes the long value.

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/serialization/InstantSerializer.java:
```diff
@@ -0,0 +1,51 @@
+import org.joda.time.Instant;
+
+/**
+ * Kafka {@link Serializer} for {@link Instant}.
```
Ditto

peay replied:
Fixed.

rangadi commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/serialization/InstantDeserializer.java:
```diff
@@ -0,0 +1,51 @@
+ *
+ * <p>This uses the same encoding as {@link org.apache.beam.sdk.coders.InstantCoder}.
+ */
+public class InstantDeserializer implements Deserializer<Instant> {
```
Why do we have this class here?

peay replied:
Some of the tests use `KafkaIO` with `Instant` keys.

rangadi commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -249,16 +330,74 @@
+    // configuration.
+    ImmutableMap.Builder<String, Object> builder = new ImmutableMap.Builder<>();
+    Map<String, Object> config = builder
+            .putAll(Read.DEFAULT_CONSUMER_PROPERTIES)
```
This would serve as an example for others to their own coders as serializers.
They wouldn't have access to `Read.DEFAULT_CONSUMER_PROPERTIES` and other builders. Can you rewrite it using only the public api?

In fact you could implement `fromCoders()` (again using only public API) and fromAvro would just invoke that.

peay replied:
OK, good point. I have added `readWithCoders` and `writeWithCoders`.

rangadi commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -249,16 +330,74 @@
+   * container.
+   */
+  @SuppressWarnings("unchecked")
+  public static <K, V> Write<K, V> toAvro(Class<K> keyClass, Class<V> valueClass) {
```
Same comment as with `fromAvor()`. Better if this only uses public API. Also you could add 'withCoderBasedSerializers()' utility method.

peay replied:
I have added `readWithCoders` and `writeWithCoders`.

rangadi commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -912,6 +1093,16 @@ public boolean start() throws IOException {
       consumerSpEL.evaluateAssign(consumer, spec.getTopicPartitions());
 
+      try {
+        keyDeserializerInstance = source.spec.getKeyDeserializer().newInstance();
```
These need to be properly closed in `close()`.

peay replied:
Fixed.

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -118,17 +122,17 @@
- *       .withValueCoder(StringUtf8Coder.of())
+ *       .withKeyDeserializer(new LongDeserializer())
+ *       .withValueDeserializer(new StringDeserializer())
+ *
```
Why? It's not the case for coders either. You supply a coder, rather than a coder class; makes sense to supply a deserializer, not deserializer class. There can be useful fields in a deserializer class in general.

rangadi replied:
Yes, Kafka interface requires class with a no-arg constructor. This is a fix for JavaDoc, actual methods ('withKeyDeserializer()`) take class. 
@peay, the top level JavaDoc will be the primary usage reference. Please verify it is correct (compiles etc).

peay replied:
Yes, fixed.

jkff commented:
We can't and shouldn't derive serializers. There's no registry for serializers like there is CoderRegistry for coders, and also the wire format of Kafka messages should be specified explicit rather than inferred implicitly.

rangadi commented:
I didn't mean 'deriving' serializers. Since user might provide coders anyway, KafkaIO could use `CoderBasedKafkaDeserializer`. I assumed `withKeyCoder()` and `withValueCoder()` methods will be removed in PR (or at least deprecated). Looks like those two are still required.

peay commented:
@jkff @rangadi thanks for the comments, I'll make a pass this weekend.

@jkff got it for CI. Maybe https://beam.apache.org/contribute/contribution-guide/ could be updated? In particular, `We recommend enabling Travis-CI continuous integration coverage on your private fork in order to easily test your changes before proposing a pull request. Go to Travis-CI, log in with your GitHub account, and enable coverage for your repository.` is a bit misleading.

jkff commented:
Huh, @rangadi I guess we could make a serializer-based Coder - i.e. reuse the serializer (which the user has to provide) to also encode/decode values for Beam. And then indeed we could remove withKeyCoder and withValueCoder methods, and the whole CoderRegistry machinery. @peay WDYT?

rangadi commented:
> reuse the serializer (which the user has to provide) to also encode/decode values for Beam. 
User does not have to provide serializer for a reader (just needs deserializer). I think it is better for KafkaIO to follow recommended practice to handle Coders.

rangadi commented:
> User does not have to provide serializer for a reader (just needs deserializer). I think it is better for KafkaIO to follow recommended practice to handle Coders.

I thought Beam would derive the coders from type parameters. In fact runners could do this at the transform level rather than individual transforms.

jkff commented:
The current PR implements inference of coders from type parameters of the Serializer/Deserializer. Due to Java's type erasure, Beam itself can't always do it automatically - otherwise users would never need to set coders anywhere (which, granted, would be great).

rangadi commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -214,18 +251,62 @@
+  @Nullable
+  public static <T> Coder<T> getOrInferCoder(CoderRegistry coderRegistry,
+                                             Class<? extends Deserializer<T>> deserializer,
+                                             Coder<T> coder) {
```
remove the third arg.

peay replied:
OK, done.

rangadi commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -214,18 +251,62 @@
+      return coder;
+    }
+
+    if (deserializer == null) {
```
checkNotNull(), no need to warn.. (it is already checked by the caller to be non null).

peay replied:
Fixed.

rangadi commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -214,18 +251,62 @@
+    }
+
+    // Get Deserializer<T>
+    Type type = deserializer.getGenericInterfaces()[0];
```
+1

rangadi replied:
@peay, this method really needs some unit tests. Could you add them to KafkaIOTest.java?

peay replied:
Excellent point, I've modified the code. 

The method already had unit tests. I have added an extra test for the case of multiple interfaces.

rangadi commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -214,18 +251,62 @@
+    // Get Deserializer<T>
+    Type type = deserializer.getGenericInterfaces()[0];
+
+    if (type instanceof ParameterizedType) {
```
What happens if we use this with `CoderBasedKafkaDeserializer` for Kafka serializer? Please include it in the unit tests.

peay commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -214,18 +251,62 @@
+      } catch (CannotProvideCoderException | ClassCastException e) {
+        LOG.warn("Could not infer coder from deserializer type {}: {}",
+                deserializer, e.getMessage());
+        e.printStackTrace();
```
Ooops, fixed.

peay commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -176,8 +210,8 @@
- *       .withKeyCoder(BigEndianLongCoder.of())
- *       .withValueCoder(StringUtf8Coder.of())
+ *       .withKeySerializer(new LongSerializer())
+ *       .withValueSerializer(new StringSerializer())
```
Yes, fixed.

peay commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -441,15 +599,35 @@ public void validate(PBegin input) {
+      CoderRegistry registry = input.getPipeline().getCoderRegistry();
+
+      // Handles unbounded source to bounded conversion if maxNumRecords or maxReadTime is set.
+      Coder<K> keyCoder = getOrInferCoder(registry, getKeyDeserializer(), getKeyCoder());
```
Not sure what you mean here.

peay commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -912,6 +1093,16 @@ public boolean start() throws IOException {
+        keyDeserializerInstance = source.spec.getKeyDeserializer().newInstance();
+        valueDeserializerInstance = source.spec.getValueDeserializer().newInstance();
+      } catch (InstantiationException | IllegalAccessException e) {
+        throw new IOException("Could not instantiate deserializers: " + e.toString());
```
Fixed.

peay commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -912,6 +1093,16 @@ public boolean start() throws IOException {
+        throw new IOException("Could not instantiate deserializers: " + e.toString());
+      }
+
+      this.keyDeserializerInstance.configure(spec.getConsumerConfig(), true);
```
Fixed.

peay commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -1033,13 +1224,12 @@ public boolean advance() throws IOException {
       // This makes it impossible for user to distinguish between zero length byte and null.
-      // Alternately, we could have a ByteArrayCoder that handles nulls, and use that for default
-      // coder.
       byte[] toDecode = bytes == null ? nullBytes : bytes;
```
OK, I have removed the function.

peay commented:
Thanks a lot for your comments. I did a pass, including:
- Renamed `getOrInferCoder` to `inferCoder`, removed `coder` argument
- Removed `deserialize` method and the null bytes hack
- Close deserializers
- Added `readWithCoders` and `writeWithCoders`
- Rewrote the Javadoc a bit
- Handle classes with multiple interfaces, added a test

With `readWithCoders` and `writeWithCoders`, manually calling `with{Key,Value}Coder` is not longer necessary in cases where inference does not work. I would still leave them around, but simply do not mention them in the main Javadoc. This means the following usages:

- you use `with{Key,Value}Deserializer` with a Kafka deserializer. This avoids an extra level of wrapping, so this is probably preferable in the simple cases.
- you use `readWithCoders` with `Coder`s. This is good for backward compatibility or if you do not want to develop a custom Kafka deserializer. I'll let you judge to what extent this collides with `Do not use coders as a general serialization or parsing mechanism` from the style guide.

Same with writing. What do you think?

Still missing:
- tests for `readWithCoders`, `writeWithCoders` (and their Avro instantiations )?

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -150,6 +154,49 @@
  *
+ * <p>Kafka provides deserializers for common types in
+ * {@link org.apache.kafka.common.serialization}. Alternatively, to use Kafka deserializers
+ * based on wrapping Beam {@link Coder}s, use {@code readWithCoders} instead of {@code read}:
```
Add a note here that this is discouraged because the wire format of a Coder is not guaranteed, so this will only work if the data is exchanged with another Beam pipeline using KafkaIO with the same coder?

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -150,6 +154,49 @@
+ *
+ * <p>The key and value {@link Coder}s for the resulting collection are inferred from the
+ * deserializer types through the {@link Coder} registry. When the registry cannot infer the
+ * coder types, either register custom {@link Coder}s in the registry, or use
```
This comment is confusing: you're still allowed to explicitly specify .withKeyCoder() and .withValueCoder(), right? I'd recommend to phrase this paragraph simply as:

In most cases, there's no need to specify key or value coders because they will be inferred from the respective deserializer. However, if that fails, you can specify them explicitly using withKeyCoder/withValueCoder().

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -175,9 +222,8 @@
- *       // set Coder for Key and Value
- *       .withKeyCoder(BigEndianLongCoder.of())
- *       .withValueCoder(StringUtf8Coder.of())
+ *       .withKeySerializer(new LongSerializer())
```
.class

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -193,11 +239,14 @@
  *      .withBootstrapServers("broker_1:9092,broker_2:9092")
  *      .withTopic("results")
- *      .withValueCoder(StringUtf8Coder.of()) // just need coder for value
+ *      .withValueSerializer(new StringSerializer()) // just need serializer for value
```
.class

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -214,18 +263,57 @@
+   * Attempt to infer a {@link Coder} by extracting the type of the deserialized-class
+   * from the deserializer argument using the {@link Coder} registry.
+   */
+  public static <T> Coder<T> inferCoder(CoderRegistry coderRegistry,
```
This shouldn't be public - better make it package-local (no access modifier) and mark `@VisibleForTesting`.

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -214,18 +263,57 @@
+    }
+
+    throw new RuntimeException("Could not extract deserializer type from "
+            + deserializer.toString());
```
.toString() is unnecessary

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -249,15 +337,90 @@
+            .setTopicPartitions(new ArrayList<TopicPartition>())
+            .setKeyCoder(keyCoder)
+            .setValueCoder(valueCoder)
+            .setKeyDeserializer((Class<Deserializer<K>>) (Class<?>) keyDeserializer.getClass())
```
Both of these are equivalent to simply CoderBasedKafkaDeserializer.class.

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -249,15 +337,90 @@
+
+    return new AutoValue_KafkaIO_Write.Builder<K, V>()
+            .setProducerConfig(config)
+            .setKeySerializer((Class<Serializer<K>>) (Class<?>) keySerializer.getClass())
```
Same

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java:
```diff
@@ -1003,15 +1229,16 @@ public boolean advance() throws IOException {
 
-          // apply user coders. might want to allow skipping records that fail to decode.
+          // apply user deserializers.
+          // might want to allow skipping records that fail to deserialize.
```
Silently skipping bad records is silent data loss. A better solution for this case is to have a sort of a "dead-letter" side output, containing raw records that could not be deserialized. That's outside the scope of this PR though.

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/serialization/CoderBasedKafkaDeserializer.java:
```diff
@@ -0,0 +1,72 @@
+ * must be stored as serialized value in the producer configuration map.
+ */
+public class CoderBasedKafkaDeserializer<T> implements Deserializer<T> {
+    @SuppressWarnings("unchecked")
```
Looks like this class uses different indentation (4 spaces) than the rest of Beam (2 spaces).

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/serialization/InstantDeserializer.java:
```diff
@@ -0,0 +1,51 @@
+ * <p>This decodes the number of milliseconds since epoch using {@link LongDeserializer}.
+ */
+public class InstantDeserializer implements Deserializer<Instant> {
+    private LongDeserializer longDeserializer = null;
```
Here too please use 2 spaces.

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/serialization/InstantSerializer.java:
```diff
@@ -0,0 +1,51 @@
+ * <p>This encodes the number of milliseconds since epoch using {@link LongSerializer}.
+ */
+public class InstantSerializer implements Serializer<Instant> {
+    private LongSerializer longSerializer = null;
```
And here

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/serialization/InstantDeserializer.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Override
+    public void configure(Map<String, ?> configs, boolean isKey) {
+        if (longDeserializer == null) {
```
Why not just initialize it at declaration? It's cheap to create, and in any realistic pipeline you're going to create it sooner or later anyway, so no performance is gained by lazy initialization.

jkff commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/serialization/InstantSerializer.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Override
+    public void configure(Map<String, ?> configs, boolean isKey) {
+        if (longSerializer == null) {
```
Same here too (initialize at declaration)

peay commented:
I am travelling until end

peay commented:
I am travelling until end of next week. If you can finish this up, that's great -- otherwise, I can make a pass on Friday. Thanks!

rangadi commented:
Thanks @peay. Are you squashing after each commit? I think it is simpler to review just the changes from last review rather than looking at the entire patch each time.

jkff commented:
Thanks, I did the cleanups and merged.
